### ISSUE TITLE
`Wmissing-prototypes` and `Wmissing-variable-declarations` should not trigger in a module

### ISSUE THREAD

**Issue by davidstone** [llvm/llvm-project#192996]
The following translation unit

```c++
export module m;

void f() {
}

int x;
```

when compiled with `clang++ -std=c++20 -Wmissing-prototypes -Wmissing-variable-declarations` gives a warning of

```console
<source>:3:6: warning: no previous prototype for function 'f' [-Wmissing-prototypes]
    3 | void f() {
      |      ^
<source>:3:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
    3 | void f() {
      | ^
      | static 
<source>:6:5: warning: no previous extern declaration for non-static variable 'x' [-Wmissing-variable-declarations]
    6 | int x;
      |     ^
<source>:6:1: note: declare 'static' if the variable is not intended to be used outside of this translation unit
    6 | int x;
      | ^
2 warnings generated.
Compiler returned: 0
```

The purpose of these warnings is to prevent violations of the one-definition rule. Modules already prevent this problem because the function is defined in the module purview, so no warning is needed in these cases.

See it live: https://godbolt.org/z/PGPsYbqfz

**Comment by PrabbyDD** [comment]
I can try this one. It seems like this mean I need to find out where the warning is and suppress/remove it?

**Comment by AaronBallman** [comment]
> The purpose of these warnings is to prevent violations of the one-definition rule.

I'm surprised either triggers in C++ code at all. `-Wmissing-prototypes` makes no sense in C++ because every function in C++ has a prototype and if you spell it wrong, you get an overload. Whereas in C you can spell a function without a prototype (pre C23) and redeclare it with a prototype by accident: https://godbolt.org/z/ja35KxKEW

GCC documents both as being C-only:
https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wmissing-prototypes
https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wmissing-variable-declarations

But Clang implements a different behavior for `-Wmissing-declarations` which is the C++-specific variant of these: https://github.com/llvm/llvm-project/issues/16660

**Comment by AaronBallman** [comment]
> I can try this one. It seems like this mean I need to find out where the warning is and suppress/remove it?

It's worth discussing what should happen, first.

1) We should figure out which `-Wmissing-declarations` scenarios we don't yet cover in C++ and implement those, but it's unclear what diagnostic groups to implement them under given the existing divergence between Clang and GCC here. Possibly fine to put them under `-Wmissing-declarations`, might need their own group though.
2) As we get coverage, we can disable `-Wmissing-prototypes` and `-Wmissing-variable-declarations` in C++ mode.

The purpose to `-Wmissing-declarations` isn't to avoid ODR violations; it's for code hygiene to ensure that external definitions in source are matched to an external declaration in a header file.

**Comment by davidstone** [comment]
The value I get out of these warnings is that I want the following rules to apply to every kind of thing, where applicable, assuming we're not in a module:

* If it's defined in a header file, it should be `inline`. This applies to functions and, since C++17, variables. Some functions are implicitly `inline` and just don't need the keyword (for instance, functions defined in the class body).
* If it's declared and defined in a source file, it should be either `static` or in the unnamed namespace. This applies to pretty much everything including class definitions, alias declarations, etc.
* Everything else should be declared (`extern` if applicable) in a header and defined in the corresponding source file. Function declarations are implicitly `extern` so they don't need the keyword.

Following this rule avoids ODR violations and maximizes chances for optimization.

In a module, none of this is necessary. Instead it just becomes everything should be either exported or not exported, which is trivially true.

**Comment by ChuanqiXu9** [comment]
@PrabbyDD FYI https://github.com/llvm/llvm-project/issues/193004

**Comment by ChuanqiXu9** [comment]
> In a module, none of this is necessary. Instead it just becomes everything should be either exported or not exported, which is trivially true.

BTW, if the decl shouldn't be accessed by any other TU (even if in the same module), it is better to declare it with static or anony namespace.

**Comment by PrabbyDD** [comment]
> [@PrabbyDD](https://github.com/PrabbyDD) FYI [#193004](https://github.com/llvm/llvm-project/issues/193004)

Hmm. I guess someone is already handling a similar issue? If thats the case I can look for something else to work on.

**Comment by ChuanqiXu9** [comment]
> > [@PrabbyDD](https://github.com/PrabbyDD) FYI [#193004](https://github.com/llvm/llvm-project/issues/193004)
> 
> Hmm. I guess someone is already handling a similar issue? If thats the case I can look for something else to work on.

He is handling another issue with similar pattern

**Comment by localspook** [comment]
> The purpose to -Wmissing-declarations isn't to avoid ODR violations; it's for code hygiene to ensure that external definitions in source are matched to an external declaration in a header file.

> If it's declared and defined in a source file, it should be either static or in the unnamed namespace. This applies to pretty much everything including class definitions, alias declarations, etc.

This is exactly what the clang-tidy check [`misc-use-internal-linkage`](https://clang.llvm.org/extra/clang-tidy/checks/misc/use-internal-linkage.html) enforces, and this issue even looks very similar to one we had with that check: #187884.

Am I understanding correctly that, essentially, we want `-Wmissing-declarations` to be equivalent to `misc-use-internal-linkage`?

**Comment by AaronBallman** [comment]
> > The purpose to -Wmissing-declarations isn't to avoid ODR violations; it's for code hygiene to ensure that external definitions in source are matched to an external declaration in a header file.
> 
> > If it's declared and defined in a source file, it should be either static or in the unnamed namespace. This applies to pretty much everything including class definitions, alias declarations, etc.
> 
> This is exactly what the clang-tidy check [`misc-use-internal-linkage`](https://clang.llvm.org/extra/clang-tidy/checks/misc/use-internal-linkage.html) enforces, and this issue even looks very similar to one we had with that check: [#187884](https://github.com/llvm/llvm-project/issues/187884).
> 
> Am I understanding correctly that, essentially, we want `-Wmissing-declarations` to be equivalent to `misc-use-internal-linkage`?

I think that might be the conclusion we're driving towards, yes. (That and not emitting `-Wmissing-prototypes` or `-Wmissing-variable-declarations` in C++ mode.)

**Comment by ChuanqiXu9** [comment]
From my reading, I feel @localspook and @AaronBallman are not on the same page. @localspook means to not emit the warning if the source is not importable  unit nor header. But @AaronBallman said we shouldn't warn these warning in C++ at all if I read correctly.

**Comment by AaronBallman** [comment]
I thought we were on the same page, but good to confirm to be sure. :-) I was reading @localspook as saying that `-Wmissing-declarations` behavior is effectively equivalent to `misc-use-internal-linkage`; situations where the tidy check warns are the same situations we'd want the compiler to warn on under `-Wmissing-declarations` (or a new flag, because https://github.com/llvm/llvm-project/issues/16660 is still an open question), and I was saying that once we surfaced that functionality in the compiler, we'd stop emitting `-Wmissing-prototypes` and `-Wmissing-variable-declarations` in C++ mode because those diagnostics don't make sense for C++ beyond what would now be covered by the new diagnostic.

**Comment by localspook** [comment]
That's how I understood you, I believe we're on the same page ^_^
